src/els_code_action_provider.erl

```diff
@@ -0,0 +1,128 @@
+%% Internal Functions
+%%==============================================================================
+
+
+%% result: (Command | CodeAction)[] | null where CodeAction is defined
+%% as follows:
+-spec code_actions(uri(), range(), code_action_context()) -> [map()].
+code_actions(Uri, Range, Context) ->
+  lager:info("code_actions: [Context=~p]", [Context]),
+  #{ <<"start">> := #{ <<"character">> := _StartCol
+                     , <<"line">>      := StartLine }
+   , <<"end">> := _End
+   } = Range,
+  #{ <<"diagnostics">> := Diagnostics } = Context,
+  %% TODO: process optional "only" field in Context
+  Actions0 = [ make_code_action(Uri, C) || C <- Diagnostics],
+
+  Actions = case application:get_env(erlang_ls, test_code_action) of
+    {ok, true} -> [ add_lines_action( Uri
+                                    , <<"Add TODO comment">>
+                                    , <<"%% TODO: something\n">>
+                                    , StartLine)]
+                   ++ Actions0;
+    _ -> Actions0
+  end,
+  lager:info("code_actions: [Actions=~p]", [Actions]),
+  Actions.
+
+
+%% TODO: either provide a literal workspaceEdit, or a command, based
+%% on the `textDocument.codeAction.codeActionLiteralSupport` client
+%% capability.
+-spec add_lines_action(uri(), binary(), binary(), number()) -> map().
+add_lines_action(Uri, Title, Lines, Before) ->
+   #{ title => Title
+    , command =>
+          els_utils:make_command( Title
+                                , <<"add-lines">>
+                                , [#{ uri => Uri
+                                    , lines => Lines
+                                    , before => Before }])
+    }.
+
+-spec make_code_action(uri(), diagnostic()) -> [code_action()].
+make_code_action(Uri, #{ <<"message">> := Message
+                       , <<"range">> := Range } = _Diagnostic) ->
+    lager:info("make_code_action: [Message=~p]", [Message]),
+    %% processing messages of the type "type type_name() undefined".
+    %% Good resource:
+%% https://arifishaq.wordpress.com/2013/11/13/playing-with-regular-expressions/
+    %% TODO: this is a hot path, precompile the re's.
+    case re:run(Message, "type (.*) undefined"
+               , [{capture, all_but_first, binary}]) of
+        {match, [UndefinedType]} ->
+            undefined_type_action(Uri, Range, UndefinedType);
+        _ -> []
+    end.
+
+
+-spec undefined_type_action(uri(), range(), binary()) -> [code_action()].
+undefined_type_action(Uri, _Range, UndefinedType) ->
+    lager:info("undefined_type_action: [UndefinedType=~p]", [UndefinedType]),
+    R = els_parser:parse(UndefinedType),
+    lager:info("undefined_type_action: [R=~p]", [R]),
+    {ok, Tokens, _} = erl_scan:string(binary_to_list(UndefinedType) ++ "."),
+    {ok, [{call, _, Name, Args}]} = erl_parse:parse_exprs(Tokens),
+    lager:info("undefined_type_action: [{Name,Args}=~p]", [{Name, Args}]),
+    MFA = case Name of
```

Comment:
We should probably move this to its own function and, if possible, use the `erl_syntax` API.

Reply:
I will move it to its own function, but I gave `erl_syntax` a quick scan and it is not clear to me what I should use.